wallet/udb/txmined.go

```diff
@@ -1445,6 +1445,160 @@ func (s *Store) addCredit(ns walletdb.ReadWriteBucket, rec *TxRecord, block *Blo
+		// else to do. Note that this means unspent ticket commitment entries
+		// are added to the index only once, at the very first time the ticket
+		// commitment is seen.
+		return nil
+	}
+
+	if index >= uint32(len(rec.MsgTx.TxOut)) {
+		return errors.E(errors.Invalid, "index should be of an existing output")
+	}
+
+	if (index % 2 != 1) {
+		return errors.E(errors.Invalid, "index should be of a ticket commitment")
+	}
+
+	if rec.TxType != stake.TxTypeSStx {
+		return errors.E(errors.Invalid, "transaction record should be of a ticket")
+	}
+
+	txOut := rec.MsgTx.TxOut[index]
+	txOutAmt, err := stake.AmountFromSStxPkScrCommitment(txOut.PkScript)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Accounting for ticket commitment %v:%d (%v) from the wallet",
+		rec.Hash, index, txOutAmt)
+
+	v = valueTicketCommitment(txOutAmt, account)
+	err = putRawTicketCommitment(ns, k, v)
+	if err != nil {
+		return err
+	}
+
+	v = valueUnspentTicketCommitment(false)
+	return putRawUnspentTicketCommitment(ns, k, v)
+}
+
+// originalTicketInfo returns the transaction hash and output count for the
+// ticket spent by the given transaction.
+//
+// spenderType MUST be either TxTypeSSGen or TxTypeSSRtx and MUST by the type of
+// the provided spenderTx, otherwise the result is undefined.
+func originalTicketInfo(spenderType stake.TxType, spenderTx *wire.MsgTx) (chainhash.Hash, uint32) {
+	if spenderType == stake.TxTypeSSGen {
+		// Votes have an additional input (stakebase) and two additional outputs
+		// (previous block hash and vote bits) so account for those.
+		return spenderTx.TxIn[1].PreviousOutPoint.Hash,
+			uint32((len(spenderTx.TxOut)-2)*2+1)
+	}
+
+	return  spenderTx.TxIn[0].PreviousOutPoint.Hash,
+		uint32(len(spenderTx.TxOut)*2+1)
+}
+
+// removeUnspentTicketCommitments deletes any outstanding commitments that are
+// controlled by this wallet and have been redeemed by the given transaction.
+//
+// rec MUST be either a vote or revocation transaction, otherwise the results
+// are undefined.
+func (s *Store) removeUnspentTicketCommitments(ns walletdb.ReadWriteBucket,
+	txType stake.TxType, tx *wire.MsgTx) error {
+
+	ticketHash, ticketOutCount := originalTicketInfo(txType, tx)
+	for i := uint32(1); i < ticketOutCount ; i+= 2 {
+		k := keyTicketCommitment(ticketHash, i)
+		if existsRawTicketCommitment(ns, k) == nil {
+			// This commitment was not tracked by the wallet, so ignore it.
+			continue
+		}
+
+		log.Debugf("Removing unspent ticket commitment %v:%d from the wallet",
+			ticketHash, i)
+
+		err := deleteRawUnspentTicketCommitment(ns, k)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// replaceTicketCommitmentUnimedSpent replaces the unminedSpent flag of all
+// unspent commitments spent by the given transcation to the given value. Note
+// that this has a side effect of creating an entry in the unspent ticket
+// commitment index if one does not yet exist.
```

Comment:
Is this note still true? If the ticket commitment for a tx hash + output index isn't already recorded, it is skipped by the loop.